lock-api/src/main/java/com/palantir/lock/client/MultiClientCommitTimestampGetter.java

```diff
@@ -0,0 +1,193 @@
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.watch.ImmutableTransactionUpdate;
+import com.palantir.lock.watch.LockWatchEventCache;
+import com.palantir.lock.watch.LockWatchVersion;
+import com.palantir.lock.watch.TransactionUpdate;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import org.immutables.value.Value;
+
+public final class MultiClientCommitTimestampGetter implements AutoCloseable {
+    private final DisruptorAutobatcher<NamespacedRequest, Long> autobatcher;
+
+    private MultiClientCommitTimestampGetter(DisruptorAutobatcher<NamespacedRequest, Long> autobatcher) {
+        this.autobatcher = autobatcher;
+    }
+
+    public static MultiClientCommitTimestampGetter create(InternalMultiClientConjureTimelockService delegate) {
+        DisruptorAutobatcher<NamespacedRequest, Long> autobatcher = Autobatchers.independent(consumer(delegate))
+                .safeLoggablePurpose("multi-client-get-commit-timestamp")
+                .build();
+        return new MultiClientCommitTimestampGetter(autobatcher);
+    }
+
+    public long getCommitTimestamp(Namespace namespace, long startTs, LockToken commitLocksToken) {
+        return AtlasFutures.getUnchecked(autobatcher.apply(ImmutableNamespacedRequest.builder()
+                .namespace(namespace)
+                .startTs(startTs)
+                .commitLocksToken(commitLocksToken)
+                .build()));
+    }
+
+    @VisibleForTesting
+    static Consumer<List<BatchElement<NamespacedRequest, Long>>> consumer(
+            InternalMultiClientConjureTimelockService delegate) {
+        return batch -> {
+            BatchStateManager batchStateManager = new BatchStateManager(getNamespaceWiseBatchStateManager(batch));
+            while (batchStateManager.hasPendingRequests()) {
+                batchStateManager.processResponse(delegate.getCommitTimestamps(batchStateManager.getRequests()));
+            }
+        };
+    }
+
+    private static Map<Namespace, NamespacedBatchStateManager> getNamespaceWiseBatchStateManager(
+            List<BatchElement<NamespacedRequest, Long>> batch) {
+        Map<Namespace, NamespacedBatchStateManager> requestMap = new HashMap<>();
+
+        for (BatchElement<NamespacedRequest, Long> elem : batch) {
+            NamespacedRequest argument = elem.argument();
+            Namespace namespace = argument.namespace();
+            NamespacedBatchStateManager namespacedBatchStateManager =
+                    requestMap.computeIfAbsent(namespace, _unused -> new NamespacedBatchStateManager(argument.cache()));
+            namespacedBatchStateManager.addRequest(elem);
+        }
```

Comment:
This function makes sense, but I'd expect to live within the BatchStateManager as opposed to here.